annorxiver_modules/corpora_comparison_helper.py

```diff
@@ -0,0 +1,411 @@
+        count_dict - the counter object to be outputted
+        file_name - the name of the file for the output to be written to
+    """
+    # Avoids memory issues
+    with open(file_name, "w") as tsvfile:
+        reader = csv.DictWriter(
+            tsvfile, 
+            fieldnames=["lemma", "pos_tag", "dep_tag", "count"], 
+            delimiter="\t"
+        )
+
+        reader.writeheader()
+        for item in tqdm_notebook(count_dict):
+            reader.writerow({
+                "lemma":item[0],
+                "pos_tag":item[1],
+                "dep_tag":item[2],
+                "count":count_dict[item]
+            })
+
+def get_term_statistics(corpus_one, corpus_two, freq_num, psudeocount=1):
+    """
+    This function is designed to perform the folllowing calculations:
+        - log likelihood of contingency table
+        - log odds ratio
+        
+    keywords:
+        corpus_one - a dataframe object with terms and counts 
+        corpus_two - a datafram object with terms and counts
+        freq_num - number of most common words to use from both corpora
+    """
+    term_list = (
+        set(
+            corpus_one
+            .sort_values("count", ascending=False)
+            .head(freq_num)
+            .lemma
+            .values
+        )
+        |
+        set(
+            corpus_two
+            .sort_values("count", ascending=False)
+            .head(freq_num)
+            .lemma
+            .values
+        )
+    )
+    
+    corpus_one_total = corpus_one['count'].sum()
+    corpus_two_total = corpus_two['count'].sum()
+    
+    term_data = []
+    for term in tqdm_notebook(term_list):
+
+        corpus_one_term_count = (
+            corpus_one.query(f"lemma=='{term}'")['count'].values[0]
+            if term in corpus_one.lemma.tolist()
+            else 0
+        )
+        
+        corpus_two_term_count = (
+            corpus_two.query(f"lemma=='{term}'")['count'].values[0]
+            if term in corpus_two.lemma.tolist()
+            else 0
+        )
+        
+        observed_contingency_table = np.array([
+            [corpus_one_term_count, corpus_two_term_count],
+            [corpus_one_total, corpus_two_total]
+        ])
+
+        # Log Likelihood
+
+        ## add psudeocount to prevent log(0)
+        observed_contingency_table += psudeocount
+
+        a, b, c, d = (
+            observed_contingency_table[0][0],
+            observed_contingency_table[0][1],
+            observed_contingency_table[1][0],
+            observed_contingency_table[1][1]
+        )
+
+        # Obtained from (Kilgarriff, 2001) - Comparing Corpora
+        LL = lambda a,b,c,d: 2*(
+            a*np.log(a) + b*np.log(b) + c*np.log(c) + d*np.log(d)
+            - (a+b)*np.log(a+b) - (a+c)*np.log(a+c) - (b+d)*np.log(b+d)
+            - (c+d)*np.log(c+d) + (a+b+c+d)*np.log(a+b+c+d)
+        )
+        log_likelihood = LL(a,b,c,d)
+
+
+        # Log Odds
+        log_ratio = float((a*d)/(b*c))
```

Comment:
Sorry I remember talking to you about the odds ratio representing the likelihood of seeing a word in one corpora versus another. But I'm getting a little lost with the total in your contingency table since usually there isn't a total... So you have (count in corpora 1/count in corpora 2)/(total count in corpora 1/total count in corpora 2)...So its the odds of seeing a word in corpora 1 vs seeing it in the total..?

Reply:
> But I'm getting a little lost with the total in your contingency table since usually there isn't a total... So you have (count in corpora 1/count in corpora 2)/(total count in corpora 1/total count in corpora 2)...So its the odds of seeing a word in corpora 1 vs seeing it in the total..?

Almost got it. The total in this case is the total number of tokens in both corpora. The odds ratio for this is calculated as follows: (token count in corpora 1/total number of tokens in corpora 1)/((token count in corpora 2/total number of tokens in corpora 2). The odds ratio depicts how likely a token is to appear in one corpus vs another. So values > 1 indicate the token is more likely to appear in corpus one rather than corpus two.